insights/core/spec_factory.py

```diff
@@ -117,7 +117,10 @@ def content(self):
                 raise
 
         if len(self._content) == 0:
-            raise ContentException("Empty (after filtering): %s" % self.path)
+            log.debug("File is empty (after filtering): %s", self.path)
```

Comment:
@xiangce It's been a while since I've worked on core stuff, but I was wondering would it be better to raise a SkipException instead here. Or would that trigger the same issue? It's already kind of hard sometimes finding out why a rule wasn't ran, now without an exception being thrown and logged. I feel like it might be even harder to determine this, but I could be thinking about it wrong and was curious about your thoughts.

Reply:
@ryan-blakley - yeah, it's a long time 😄. I think you meant `SkipComponent` (the SkipException was removed). Raising a `SkipComponent`, in most cases, it won't cause any issues. However, in our experience of rules development, it occasionally happens that, for example, it is expected that a particular keyword does not appear in a file, or that a particular directive is not explicitly configured in a "fitlterable" configuration file, but the application sets default values for those unconfigured directives, etc. In these cases, if we `Skip` these Specs in the core, the condition that depends on the Specs/Parsers will be skipped too per the dependency graph. For example:
```
@condition(Specs.A_filterable_Spec):
def no_such_line_in_file(spec):
    if "such_line" in spec.lines:
        return False
    return True

@rule(..., no_such_line_in_file, ...) 
def report(...)
```
If the above `A_filterable_Spec` is `Skipped` by the core, the condition `no_such_line_in_file` will never be triggered and hence the `report` will also be skipped.


Although we have workarounds to avoid this situation, or implement it in a `correct` way if we know the background that the spec will be Skipped when it's filtered empty, it may not mean that all rule developers are aware of this. If still implementing the `conditions` in the traditional style, we may get false positives.

Reply:
Sorry, yeah I meant SkipComponent, I should have remembered that since I removed the SkipException me--. Ahhh gotcha I didn't realize the ContentException raise was recently added, thanks for the explanation, everything looks good then to me.

Reply:
Thanks @ryan-blakley